### PR TITLE
[5.2] Cast incrementing IDs to Integer by default

### PR DESCRIPTION
**Pull request by mattstauffer**
If a model is incrementing, set a default to cast the primary key to type `integer`.

This fixes an inconsistency where an instance created with `Model::create` would have an integer ID, but model instances fetched from the database would have a string ID.

Written while pairing with @adamwathan.

Also fixes another test which broke when implementing this--it was setting the primary key of an incrementing table to the value `'foo'`, which is invalid. Updated the test to use a non-incrementing model.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -2785,7 +2785,23 @@ protected function mutateAttributeForArray($key, $value)
...
+    /**
+     * Get the casts array.
+     *
+     * @return array
+     */
+    protected function getCasts()
+    {
+        if ($this->incrementing) {
+            return array_merge([
+                $this->getKeyName() => 'integer',
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: I'd rather we used the more conventional `int` here.

**Comment by mattstauffer** [comment]
@GrahamCampbell Fixed--thanks!

**Comment by JosephSilber** [comment]
Do we really need this in every call to `getCasts`?

Why not do it once in the constructor?

**Comment by adamwathan** [comment]
@JosephSilber `getDates` works the same way, so I think either both should be done at construction or on the fly.

**Comment by JosephSilber** [comment]
@adamwathan now that we can do dates in the casts, we should be doing all of that in the constructor.

**Comment by GrahamCampbell** [comment]
> getDates works the same way, so I think either both should be done at construction or on the fly.

Maybe we should change them both in 5.2 in the interest of efficiency?

**Comment by GrahamCampbell** [comment]
Maybe leave that to another PR though, to be discussed after this is either rejected or accepted.

**Comment by taylorotwell** [comment]
I agree we'll probably need to do this one time. How do we want to handle that?

**Comment by sherbrow** [comment]
This was unexpected.
Before: `incrementing` only meant that the insert would trigger a get to set the newly generated value in the model, and in that case where it was not a integer, the value would not change
After: `incrementing` means integer casting

It breaks previously non-integer primary keys when `incrementing` was let to the default `true` value.
Not the biggest change, but still worth mentioning (and `incrementing` seems not to be documented before that breaking change in 5.2)

**Comment by adamwathan** [comment]
@sherbrow Are there any cases where would you want `$incrementing = true` with a non-integer key? Or is it more a matter of not doing the cast if someone is using a non-integer key and forget to disable `$incrementing`?

**Comment by sherbrow** [comment]
@adamwathan It's definitely the case where one would forget to disable `$incrementing`, as stated, I don't think it was in the docs before 5.2.
It appears in the upgrade guide, so it's probably my fault not to take that into account in the first place.

**Pull request merged**